upgrade/src/main/java/org/odk/collect/upgrade/AppUpgrader.kt

```diff
@@ -0,0 +1,50 @@
+package org.odk.collect.upgrade
+
+import android.app.Application
+import org.odk.collect.shared.Settings
+
+class AppUpgrader(
+    private val settings: Settings,
+    private val launchState: LaunchState,
+    private val upgrades: List<Upgrade>
+) {
+    constructor(
+        key: String,
+        settings: Settings,
+        versionCode: Int,
+        installDetector: InstallDetector,
+        upgrades: List<Upgrade>
+    ) : this(
+        settings,
+        VersionCodeLaunchState(key, settings, versionCode, installDetector),
+        upgrades
+    )
+
+    /**
+     * Runs the list of passed [Upgrade] implementations in order if this is the first launch of a
+     * new version of the app (an "app upgrade"). This should be called in (or from somewhere called
+     * from) [Application.onCreate].
+     */
+    fun upgradeIfNeeded() {
+        if (launchState.isUpgradedFirstLaunch()) {
+            upgrades.forEach {
+                val key = it.key()
+
+                if (key == null) {
+                    it.run()
+                } else if (!settings.getBoolean(key)) {
+                    it.run()
+                    settings.save(key, true)
+                }
+            }
+        } else {
+            upgrades.forEach {
```

Comment:
Doing this protects us from running into the problems we ran into with `ExistingProjectMigrator` with any other "run once" upgrades we add in the future. We don't need it at the moment but I wanted to include the functionality now while we're thinking about it as it'll be easy to forget about later.

Reply:
Now I understand it especially that you added this comment but it would be good to have a short comment in that class as well because it is really confusing and I think without it  we might want to remove this code next time we touch this class because at first glance it seem s like fomenting we do not need.